### PR TITLE
Added conv2d batching rule

### PR DESCRIPTION
**Pull request by Chillee**
Also fixes https://github.com/zou3519/functorch/issues/8

Currently doesn't support the case where bias is batched but neither the input/weights are batched.

### CODE REVIEW

File: codegen/codegen_outofplacebatching.py [hunk 1]
```diff
@@ -118,7 +137,7 @@ def parse_args(args_t):
         result.append((arg[:split_idx].strip(), arg[split_idx:].strip()))
     return tuple(result)
 
-def get_signatures(path='build/aten/src/ATen/RegistrationDeclarations.h'):
+def get_signatures(path='/home/chilli/fb/pytorch/build/aten/src/ATen/RegistrationDeclarations.h'):
```
Comment (Chillee) [inline codegen/codegen_outofplacebatching.py:140]: oops.

File: test/test_vmap.py [hunk 2]
```diff
@@ -2327,6 +2359,15 @@ def test_no_random_op_support(self):
                                         'vmap: We do not yet support calling random operations'):
                 vmap(op)(*args)
 
+    def test_conv2d(self):
+        mod = torch.nn.Conv2d(4, 8, 3)
+        arg_values = [torch.randn(2, 4, 10, 15), mod.weight, None]
```
Comment (zou3519) [inline test/test_vmap.py:2364]: We should add a test where bias isn't none

**Review by zou3519 — COMMENTED** [review]
Codegen LGTM.

The batched bias logic doesn't seem correct. Also, it doesn't look like conv2d is enabled. The batching rule otherwise looks good, added some small suggestions.

File: functorch/csrc/BatchRulesConv.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    auto bias_tensor = new_bias->movedim(*bias_bdim, *out_dim);
+    return {std::get<0>(result) + *new_bias, out_dim};
+  } else {
+    return result;
+  }
+}
+
+
+TORCH_LIBRARY_IMPL(aten, FT_BATCHED_KEY, m) {
+  // VMAP_SUPPORT("conv2d", conv2d_batching_rule);
```
[the remark this answers is not in the document]
Comment (Chillee) [reply functorch/csrc/BatchRulesConv.cpp:80]: ah yes - I had it disabled to check some runtime stuff.

File: functorch/csrc/BatchRulesConv.cpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  shape.insert(shape.begin() + src + 1, size2);
+  return at::reshape(x, shape);
+}
+
+// Does not support batch_group_count (needed for convolution backwards)
+std::tuple<Tensor,optional<int64_t>>
+conv2d_batching_rule(const Tensor& lhs, optional<int64_t> lhs_bdim, const Tensor& rhs, optional<int64_t> rhs_bdim, const optional<Tensor>& bias, optional<int64_t> bias_bdim, IntArrayRef stride, IntArrayRef padding, IntArrayRef dilation, int64_t groups) {
+  std::vector<int64_t> lhs_spec = {0,1,2,3};
+  std::vector<int64_t> rhs_spec = {0,1,2,3};
+  std::vector<int64_t> out_spec = {0,1,2,3};
```
[the remark this answers is not in the document]
Comment (Chillee) [reply functorch/csrc/BatchRulesConv.cpp:27]: I feel like I should be able to get rid of these entirely... but I wasn't totally sure (so I didn't).

[the remark this answers is not in the document]
Comment (zou3519) [reply functorch/csrc/BatchRulesConv.cpp:27]: Yeah, you definitely can just hardcode these values in because people only really use NCHW convolutions. I don't understand how MemoryFormat in PyTorch works though and how this would fit into that

File: functorch/csrc/BatchRulesConv.cpp [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    auto new_w = reshape_axis_into(*rhs_bdim, rhs_spec[0], rhs);
+    auto out = at::conv2d(new_x, new_w, new_bias, stride, padding, dilation, groups);
+    out = reshape_axis_outof(out_spec[1], lhs.sizes()[*lhs_bdim], out);
+    result = {out, out_spec[1]};
+  } else {
+    result = {at::conv2d(lhs, rhs, new_bias, stride, padding, dilation, groups), nullopt};
+  }
+  if (new_bias) {
+    auto out_dim = std::get<1>(result);
+    TORCH_INTERNAL_ASSERT(out_dim, "Batching bias without weights not currently implemented");
```
[the remark this answers is not in the document]
Comment (Chillee) [reply functorch/csrc/BatchRulesConv.cpp:70]: yeah.

File: functorch/csrc/BatchRulesConv.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Chillee) [reply functorch/csrc/BatchRulesConv.cpp:27]: The reason I left this here is because I'm not sure if we need to do something different for say, channels last convolutions?

[the remark this answers is not in the document]
Comment (zou3519) [reply functorch/csrc/BatchRulesConv.cpp]: I'm not sure either. The best way to find out is probably to write a test. Our ChannelsLast tensors have dimensions in NCHW order, but the strides have NHWC layout. My hypothesis is that channels last convolution might turn into a channels first convolution when it goes through this batching rule which would be a performance problem to solve in the future

File: functorch/csrc/BatchRulesConv.cpp [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+Tensor reshape_axis_outof(int64_t src, int64_t size1, const Tensor &x) {
+  VmapDimVector shape(x.sizes().begin(), x.sizes().end());
+  TORCH_INTERNAL_ASSERT(shape[src] % size1 == 0);
+  int64_t size2 = shape[src] / size1;
+  shape[src] = size1;
+  shape.insert(shape.begin() + src + 1, size2);
+  return at::reshape(x, shape);
+}
```
Comment (zou3519) [inline functorch/csrc/BatchRulesConv.cpp:20]: Btw, I needed these for something else so I added them in https://github.com/zou3519/functorch/commit/adc32e01d459bcb22ab2a782be112b4e67645f21 (and made it so that they can accept negative dims)

File: functorch/csrc/BatchRulesConv.cpp [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    out = reshape_axis_outof(out_spec[1], lhs.sizes()[*lhs_bdim], out);
+    result = {out, out_spec[1]};
+  } else {
+    result = {at::conv2d(lhs, rhs, new_bias, stride, padding, dilation, groups), nullopt};
+  }
+  if (new_bias) {
+    auto out_dim = std::get<1>(result);
+    TORCH_INTERNAL_ASSERT(out_dim, "Batching bias without weights not currently implemented");
+    auto bias_tensor = new_bias->movedim(*bias_bdim, *out_dim);
+    return {std::get<0>(result) + *new_bias, out_dim};
```
[the remark this answers is not in the document]
Comment (Chillee) [reply functorch/csrc/BatchRulesConv.cpp:72]: Sadly, it's even more complicated than this. It's possible for the weight to be batched but the bias isn't. Which causes problems.

**Comment by Chillee** [comment]
Bottlenecked on this: https://github.com/pytorch/pytorch/pull/58283

File: functorch/csrc/BatchRulesConv.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Chillee) [reply functorch/csrc/BatchRulesConv.cpp:12]: Decided to switch back to vectors, just since it'll probably be easier to extend to general conv.

[the remark this answers is not in the document]
Comment (zou3519) [reply functorch/csrc/BatchRulesConv.cpp:12]: nit: std::vector<int64_t> -> VmapDimVector / DimVector can avoid dynamic allocations

File: functorch/csrc/BatchingMetaprogramming.h [hunk 8]
```diff
@@ -39,6 +39,18 @@ template <class Next, class Tail>
...
+  using type = Tail;
+};
+template <class Next, class Tail>
+struct IfFirstIsTensorAndSecondisBatchDimThenTailElseNext<const optional<Tensor>&, optional<int64_t>, Next, Tail> {
+  using type = Tail;
+};
+template <class Next, class Tail>
+struct IfFirstIsTensorAndSecondisBatchDimThenTailElseNext<optional<Tensor>&, optional<int64_t>, Next, Tail> {
+  using type = Tail;
+};
```
Comment (zou3519) [inline functorch/csrc/BatchingMetaprogramming.h:53]: Lol there should be some way to metaprogram this to be one statement... but let's not worry about that for now

File: test/test_vmap.py [hunk 9]
```diff
@@ -2403,6 +2436,17 @@ def test_no_random_op_support(self):
                                         'vmap: We do not yet support calling random operations'):
                 vmap(op)(*args)
 
+    def test_conv2d(self):
```
Comment (zou3519) [inline test/test_vmap.py:2439]: Can we add some more tests to convince ourselves that this works? 
From code reading everything looks fine to me, but it would be nice to make sure that different kernel sizes, strides, padding, dilation, and groups all work out

**Review by zou3519 — APPROVED** [review]

File: test/test_vmap.py (hunk 9 above)
Comment (zou3519) [reply test/test_vmap.py:2439]: The torchvision test in this file (resnet18 I think?) should exercise a lot of the cases already
